File: java/Components/Strata.Diagnostic.Core/src/strata/diagnostic/core/repository/IDiagnosticRepository.java
//////////////////////////////////////////////////////////////////////////////
// IDiagnosticRepository.java
//////////////////////////////////////////////////////////////////////////////

package strata.diagnostic.core.repository;

import strata.diagnostic.core.common.IDiagnostic;

import java.util.Collection;
import java.util.Optional;
import java.util.function.Predicate;

public
interface IDiagnosticRepository
{
    IDiagnosticRepository
    insert(IDiagnostic diagnostic);

    IDiagnosticRepository
    remove(IDiagnostic diagnostic);

    IDiagnosticRepository
    clear();

    Optional<IDiagnostic>
    getUnique(String diagnosticName);

    Collection<IDiagnostic>
    getMatching(Predicate<IDiagnostic> predicate);

    Collection<IDiagnostic>
    getAll();

    boolean
    hasUnique(String diagnosticName);

    boolean
    hasMatching(Predicate<IDiagnostic> predicate);
}

//////////////////////////////////////////////////////////////////////////////

File: csharp/Components/Strata.Diagnostic.Core/Repository/IDiagnosticRepository.cs
//  ##########################################################################
//  # File Name: IDiagnosticRepository.cs
//  ##########################################################################

using System;
using System.Collections.Generic;
using Strata.Diagnostic.Core.Common;

namespace Strata.Diagnostic.Core.Repository
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// </summary>
    ///  
    public
    interface IDiagnosticRepository
    {
        IDiagnosticRepository
        Insert(IDiagnostic diagnostic);

        IDiagnosticRepository
        Remove(IDiagnostic diagnostic);

        IDiagnosticRepository
        Clear();

        IDiagnostic
        GetUnique(string diagnosticName);

        ICollection<IDiagnostic> 
        GetMatching(Func<IDiagnostic,bool> predicate);

        ICollection<IDiagnostic>
        GetAll();

        bool
        HasUnique(string diagnosticName);

        bool
        HasMatching(Func<IDiagnostic,bool> predicate);
    }
}

//  ##########################################################################
